app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,197 @@
+    else
+      raise ArgumentError, "Invalid entry specification: #{entry}"
     end
+    hierarchy
+  rescue ActiveRecord::RecordNotFound => exception
+    Rails.logger.warn "#{self.class} skipping object, maybe something else already deleted it: #{exception.message}"
+    []
+  rescue NameError
+    # we would be here in case of a bad crafted hierarchy entry or something else unrecoverable
+    # for example NoMethodError on nil class where retrying is pointless but logging the error is needed
+    raise DoNotRetryError, "seems like unexpectedly broken delete hierarchy entry: #{entry}"
+  end
 
-    attr_reader :main_object, :reflection, :caller_worker_hierarchy
-
-    private
-
-    def destroy_has_many_association
-      main_object.public_send("#{reflection.name.to_s.singularize}_ids").each do |associated_object_id|
-        associated_object = reflection.class_name.constantize.new
-        associated_object.id = associated_object_id
-        delete_associated_object_later(associated_object)
+  # @return a single associated object for deletion or nil if non in the association
+  def handle_association(ar_object, association, hierarchy_association_string)
+    reflection = ar_object.class.reflect_on_association(association)
+
+    case reflection.macro
+    when :has_many
+      # here we keep original hierarchy entry if we still find an associated object
+      dependent = ar_object.public_send(association).take
+      if dependent
+        dependent.destroyed_by_association = reflection
+        [hierarchy_association_string, *hierarchy_entries_for(dependent)]
+      else
+        []
       end
-    rescue ActiveRecord::UnknownPrimaryKey => exception
-      Rails.logger.info "DeleteObjectHierarchyWorker#perform raised #{exception.class} with message #{exception.message}"
+    when :has_one
+      # maximum of one associated so we never keep the original hierarchy entry
+      dependent = ar_object.public_send(association)
+      dependent.destroyed_by_association = reflection if dependent
+      hierarchy_entries_for dependent
+    else
+      raise ArgumentError, "Cannot handle association #{ar_object}:#{association} type #{reflection.macro}"
     end
+  end
 
-    def destroy_has_one_association
-      associated_object = main_object.public_send(reflection.name)
-      delete_associated_object_later(associated_object)
-    end
+  # previously an example invocation could be with
+  # {"_aj_globalid": "gid://system/ProxyRule/1248934"},
+  # [
+  #   "Hierarchy-Service-2555418046474"
+  #   "Hierarchy-Proxy-392685"
+  #   "Hierarchy-ProxyRule-1248934"
+  # ],
+  # "destroy"
+  #
+  # While it could be called only with the first argument for the root object to be removed.
+  def compatibility(object, caller_worker_hierarchy = [], _background_destroy_method = 'destroy')
+    # maybe requeue first object from hierarchy would be adequate and uniqueness should deduplicate jobs
```

Comment:
This comment looks like a suggestion for future, but... isn't the code already doing what the comment says? :upside_down_face: 

On the other hand...  this seems to be kind of a "safety" mecanism to ensure we don't miss anything, but I guess it shouldn't be necessary?.. 

The "root" object does not have hierarchy, and it is supposed to be executed last. So in theory we could ignore all jobs with hierarchy array, and only convert the ones without - the "roots" of deletion?

But well, as this is something temporary, and uniqueness gem seems to be doing its job, probably it's not a big deal.

Reply:
This comment looks like a conflict resolution artifact. Need to update it.

Reply:
wrt logic, I don't think the job with only `object` without hierarchy will be alive for long. Once it is scheduled into a queue executor, it is gone and only the batches it scheduled remain. In such case we don't want to abandon the operation. If I'm missing something or don't understand the question, let me know.

Reply:
Well, I can't seem to find the specific logic in the (old) code, but I recall that the "root" job (with a single object and no hierarchy) was always floating around, and did not disappear.

I might be wrong of course... 

But I think it's not like it is used for scheduling other jobs and dies... something keeps it in the queue.

Reply:
Sidekiq can't keep long running jobs. I didn't really look but I can't also see how it would be possible to stay. After it is once performed, then it should finish and be gone, unless it eeps rescheduling but it also doesn't make any sense to be rescheduled and do the same thing multiple times.

Reply:
I think it might be happening here: https://github.com/3scale/porta/blob/5cda9b32bdc32afe5a2fee40e83a48045c7b8580/app/workers/delete_object_hierarchy_worker.rb#L73

So, when the original (no hierarchy) job is scheduled initially, the "child" jobs are enqueued, and the original is scheduled in 5 mins.

Now, I don't quite understand how it works after that... When the original jobs pops in the queue, what happens next? In practice I think it gets scheduled again for later. By looking at the code it looks as if the whole association chain of jobs would be queued again - but I think this is not what happens.

Anyway, not that it matters... I guess handling each job in the queue is safer. So, all good.

Reply:
Yeah, I wouldn't change. Perhaps when all batches of batches completed, then we run the `on_success` callback which would delete the object itself. But then if things were not fixed within 5 minutes, it appears to me that we would schedule what is left alive duplicating existing jobs and may explain some extrapolation of jobs what we were seeing... really hard to reason about all these jobs, retries, batches and callbacks.